eth/downloader/beaconsync.go

```diff
@@ -266,8 +266,17 @@ func (d *Downloader) fetchBeaconHeaders(from uint64) error {
 		)
 		for i := 0; i < maxHeadersProcess && from <= head.Number.Uint64(); i++ {
 			header := d.skeleton.Header(from)
+
+			// The header is not found in skeleton space, try to resolve the header
+			// reversely via skeleton tail. Note we can't retrieve the header by
+			// number directly, since there is no guarantee it's the header we want.
+			if header == nil {
+				header = d.getAncestorHeader(from, tail)
```

Comment:
I don't think this is a good idea. The new `getAncestorHeader` is a linear scanning from `tail` down to `from`. Given that we're in a loop already, that would be an O(n^2) blowup.

However, I'm unsure I understand why this lookup was introduced in the first place. When filling the chain from the beacon headers, why are we looking at the chain? IMHO if something's missing from the beacon chain, we should abort sync, not "sync from ourselves".

Reply:
It was introduced here, but I don't understand why https://github.com/ethereum/go-ethereum/pull/24610/files#diff-8cb29165929b2bd4c8f7c019a95cc8a69d3e36320703308f5c835fa97bdd8f31R270

Reply:
I guess the use case was that if I just switched to merge transition then I may not have 64 headers yet, so the pivot is pre-merge (and pre-skeleton) and thus feeding pivot and  post-pivot headers needs a local lookup. Still feels kind of wonky, very weird special case that will keep haunting us.

Reply:
Btw, with the scanning thing, we're scanning over and over and over again the exact same header range. We could have stopped after one scan and collected all the needed headers.

Also IMHO if this case should handle some very short (transition length) range, we should put some cap on the iteration count, otherwise it's asking for trouble if someone manages to iterate all the way back to the genesis somehow.

Reply:
The reason for adding lookup in local chain is exactly what you said, in this special scenario we need the feed the header from local. But this scenario is kind of common right now, we have a bunch of panics recently are all relevant with this scenario.

And also it's true that the scanning range should be short, we can limit the scanning length to be 64 to avoid the bad situation.

Reply:
Yes, I think adding a 64 limit should be welcome as that should cover the pivot leaking into the legacy sync part, but should still prevent any weird attacks. Perhaps let's add an error log if we can't find the needed header within the `fsMinFullBlocks` threshold and return an error + tear down sync to avoid the panics.

Reply:
And perhaps I would also add a warning log for the normal case - just for a few months before/after the merge - so we can see if/when it's happening and see if the bug is indeed just a transition issue or if it appears somewhere else too that can be fixed.

Reply:
All pushed, please take another look